javascript/ql/test/query-tests/Performance/ReDoS/tst.js

```diff
@@ -90,3 +90,171 @@ var bad17 = new RegExp('(a|aa?)*b');
+var bad18 = /(([^]|[^a])*)"/;
+
+// GOOD - there is no witness in the end that could cause the regexp to not match
+var bad19 = /([^"']+)*/g;
+
+// NOT GOOD
+var bad20 = /((.|[^a])*)"/;
+
+// GOOD
+var good10 = /((a|[^a])*)"/;
+
+// NOT GOOD
+var bad21 = /((b|[^a])*)"/;
+
+// NOT GOOD
+var bad22 = /((G|[^a])*)"/;
+
+// NOT GOOD
+var bad23 = /(([0-9]|[^a])*)"/;
+
+// NOT GOOD
+var bad24 = /(?:=(?:([!#\$%&'\*\+\-\.\^_`\|~0-9A-Za-z]+)|"((?:\\[\x00-\x7f]|[^\x00-\x08\x0a-\x1f\x7f"])*)"))?/;
+
+// NOT GOOD
+var bad25 = /"((?:\\[\x00-\x7f]|[^\x00-\x08\x0a-\x1f\x7f"])*)"/;
+
+// GOOD
+var bad26 = /"((?:\\[\x00-\x7f]|[^\x00-\x08\x0a-\x1f\x7f"\\])*)"/;
+
+// NOT GOOD
+var bad27 = /(([a-z]|[d-h])*)"/;
+
+// NOT GOOD
+var bad27 = /(([^a-z]|[^0-9])*)"/;
+
+// NOT GOOD
+var bad28 = /((\d|[0-9])*)"/;
+
+// NOT GOOD
+var bad29 = /((\s|\s)*)"/;
+
+// NOT GOOD
+var bad30 = /((\w|G)*)"/;
+
+// GOOD
+var good11 = /((\s|\d)*)"/;
+
+// NOT GOOD
+var bad31 = /((\d|\w)*)"/;
+
+// NOT GOOD
+var bad32 = /((\d|5)*)"/;
+
+// NOT GOOD
+var bad33 = /((\s|[\f])*)"/;
+
+// NOT GOOD
+var bad34 = /((\s|[\v]|\\v)*)"/;
+
+// NOT GOOD
+var bad35 = /((\f|[\f])*)"/;
+
+// NOT GOOD
+var bad36 = /((\W|\D)*)"/;
+
+// NOT GOOD
+var bad37 = /((\S|\w)*)"/;
+
+// NOT GOOD
+var bad38 = /((\S|[\w])*)"/;
+
+// NOT GOOD
+var bad39 = /((1s|[\da-z])*)"/;
+
+// NOT GOOD
+var bad40 = /((0|[\d])*)"/;
+
+// NOT GOOD
+var bad41 = /(([\d]+)*)"/;
+
+// GOOD - there is no witness in the end that could cause the regexp to not match
+var good12 = /(\d+(X\d+)?)+/;
+
+// GOOD - there is no witness in the end that could cause the regexp to not match
+var good13 = /([0-9]+(X[0-9]*)?)*/;
+
+// GOOD - but still flagged (always matches something)
+var good15 = /^([^>]+)*(>|$)/;
+
+// NOT GOOD
+var bad43 = /^([^>a]+)*(>|$)/;
+
+// NOT GOOD
+var bad44 = /(\n\s*)+$/;
+
+// NOT GOOD
+var bad45 = /^(?:\s+|#.*|\(\?#[^)]*\))*(?:[?*+]|{\d+(?:,\d*)?})/;
+
+// NOT GOOD
+var bad46 = /\{\[\s*([a-zA-Z]+)\(([a-zA-Z]+)\)((\s*([a-zA-Z]+)\: ?([ a-zA-Z{}]+),?)+)*\s*\]\}/;
+
+// NOT GOOD
+var bad47 = /(a+|b+|c+)*c/;
+
+// NOT GOOD
+var bad48 = /(((a+a?)*)+b+)/;
+
+// NOT GOOD
+var bad49 = /(a+)+bbbb/;
+
+// GOOD
+var good16 = /(a+)+aaaaa*a+/;
+
+// NOT GOOD
+var bad50 = /(a+)+aaaaa$/;
+
+// GOOD
+var good17 = /(\n+)+\n\n/;
+
+// NOT GOOD
+var bad51 = /(\n+)+\n\n$/;
+
+// NOT GOOD
+var bad52 = /([^X]+)*$/;
+
+// NOT GOOD
+var bad53 = /(([^X]b)+)*$/;
+
+// GOOD
+var good18 = /(([^X]b)+)*($|[^X]b)/;
+
+// NOT GOOD
+var bad54 = /(([^X]b)+)*($|[^X]c)/;
+
+// GOOD
+var good19 = /(.*,)+.+/;
+
+// GOOD
+var good20 = /((ab)+)*ababab/;
+
+// GOOD
+var good21 = /((ab)+)*abab(ab)*(ab)+/;
+
+// GOOD
+var good22 = /((ab)+)*/;
+
+// NOT GOOD
+var bad55 = /((ab)+)*$/;
+
+// GOOD
+var good23 = /((ab)+)*[a1][b1][a2][b2][a3][b3]/;
+
+// NOT GOOD
+var bad56 = /([\n\s]+)*(.)/;
+
+// GOOD - any witness passes through the accept state.
+var good24 = /(A*A*X)*/;
+
+// GOOD - but still flagged (always matches something)
+var good25 = /^([^>]+)*(>|$)/;
```

Comment:
> it seems the problem is that we don't model anchors at all 

That is also my conclusion, but I haven't been able to solve that in a nice way.

I think your suggestion will cause FNs. It will treat `/(a*)*$/` (which has catastrophic backtracking) much the same as `/(a*)*/` (which does not, because of greedy matching).

Reply:
I haven't been able to come up with any good solution to `$`.

Reply:
I wrote some notes about anchors in the other thread. Let's close this one.